File: CS/CustomizeExportAction.Module.Win/CustomizeWinExportController.cs
using System;
using System.ComponentModel;
using System.Collections.Generic;
using System.Diagnostics;
using System.Text;
using System.IO;
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.Actions;
using DevExpress.Persistent.Base;

using DevExpress.ExpressApp.Win.SystemModule;
using DevExpress.ExpressApp.SystemModule;
using DevExpress.ExpressApp.Win.Editors;
using DevExpress.ExpressApp.Win;
using System.Windows.Forms;
using DevExpress.XtraPrinting;

namespace CustomizeExportAction.Module.Win {
public partial class CustomizeWinExportController : ViewController {
    public CustomizeWinExportController() {
        InitializeComponent();
        RegisterActions(components);
        this.TargetViewType = ViewType.ListView;
    }
    private WinExportController exportController;
    protected override void OnActivated() {
        base.OnActivated();
        exportController = Frame.GetController<WinExportController>();
        exportController.Exporting += exportController_Exporting;
        exportController.CustomExport += exportController_CustomExport;
        exportController.Exported += exportController_Exported;
    }

    void exportController_CustomExport(object sender, CustomExportEventArgs e) {
        if (e.ExportType == PrintingSystemCommand.ExportXls) {
            GridListEditor gridListEditor = ((DevExpress.ExpressApp.ListView)View).Editor as GridListEditor;
            IPrintable printable = ((IExportableEditor)gridListEditor).Printable;
            XlsExportOptions options = new XlsExportOptions();
            options.SheetName = gridListEditor.Name;
            options.ShowGridLines = true;
            ComponentExporter exporter = new ComponentExporter(printable);
            exporter.Export(ExportTarget.Xls, e.Stream, options);
            e.Handled = true;
        }
    }
    void exportController_Exporting(object sender, CustomExportEventArgs e) {
        GridListEditor gridListEditor =
            ((DevExpress.ExpressApp.ListView)View).Editor as GridListEditor;
        if (gridListEditor != null) {
            XafGridView gridView = gridListEditor.GridView;
            if (HasCollapsedGroups(gridView)) {
                string message = 
                    "There are collapsed groups in the grid. Expand all groups in the exported file?";
                gridView.OptionsPrint.ExpandAllGroups =
                    WinApplication.Messaging.GetUserChoice(message, GetMessageBoxCaption(),
                    MessageBoxButtons.YesNo)
                    == DialogResult.Yes;
            }
        }
    }
    private bool HasCollapsedGroups(XafGridView gridView) {
        if (gridView.GroupCount > 0) {
            int rowHandle = -1;
            while (gridView.IsValidRowHandle(rowHandle)) {
                if (!gridView.GetRowExpanded(rowHandle)) return true;
                rowHandle--;
            }
        }
        return false;
    }
    private string GetMessageBoxCaption() {
        return String.Format(
            "{0} {1}", exportController.ExportAction.Caption,
                exportController.ExportAction.SelectedItem);
    }
    void exportController_Exported(object sender, CustomExportEventArgs e) {
        if (e.Stream is FileStream) {
            string fileName = ((FileStream)e.Stream).Name;
            if (File.Exists(fileName)) {
                e.Stream.Close();
                if (WinApplication.Messaging.GetUserChoice("Open the exported file?",
                    GetMessageBoxCaption(), MessageBoxButtons.YesNo) == DialogResult.Yes)
                    Process.Start(fileName);
            }
        }
    }
    protected override void OnDeactivated() {
        exportController.Exporting -= exportController_Exporting;
        exportController.CustomExport -= exportController_CustomExport;
        exportController.Exported -= exportController_Exported;
        base.OnDeactivated();
    }
}
}

File: VB/CustomizeExportAction.Module.Win/CustomizeWinExportController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Collections.Generic
Imports System.Diagnostics
Imports System.Text
Imports System.IO
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.Actions
Imports DevExpress.Persistent.Base

Imports DevExpress.ExpressApp.Win.SystemModule
Imports DevExpress.ExpressApp.SystemModule
Imports DevExpress.ExpressApp.Win.Editors
Imports DevExpress.ExpressApp.Win
Imports System.Windows.Forms
Imports DevExpress.XtraPrinting

Namespace CustomizeExportAction.Module.Win
Partial Public Class CustomizeWinExportController
    Inherits ViewController
    Public Sub New()
        InitializeComponent()
        RegisterActions(components)
        Me.TargetViewType = ViewType.ListView
    End Sub
    Private exportController As WinExportController
    Protected Overrides Overloads Sub OnActivated()
        MyBase.OnActivated()
        exportController = Frame.GetController(Of WinExportController)()
        AddHandler exportController.Exporting, AddressOf exportController_Exporting
        AddHandler exportController.CustomExport, AddressOf exportController_CustomExport
        AddHandler exportController.Exported, AddressOf exportController_Exported
    End Sub

    Private Sub exportController_CustomExport(ByVal sender As Object, ByVal e As CustomExportEventArgs)
        If e.ExportType = PrintingSystemCommand.ExportXls Then
            Dim gridListEditor As GridListEditor = TryCast((CType(View, DevExpress.ExpressApp.ListView)).Editor, GridListEditor)
            Dim printable As IPrintable = (CType(gridListEditor, IExportableEditor)).Printable
            Dim options As New XlsExportOptions()
            options.SheetName = gridListEditor.Name
            options.ShowGridLines = True
            Dim exporter As New ComponentExporter(printable)
            exporter.Export(ExportTarget.Xls, e.Stream, options)
            e.Handled = True
        End If
    End Sub
    Private Sub exportController_Exporting(ByVal sender As Object, ByVal e As CustomExportEventArgs)
        Dim gridListEditor As GridListEditor = TryCast((CType(View, DevExpress.ExpressApp.ListView)).Editor, GridListEditor)
        If gridListEditor IsNot Nothing Then
            Dim gridView As XafGridView = gridListEditor.GridView
            If HasCollapsedGroups(gridView) Then
                Dim message As String = "There are collapsed groups in the grid. Expand all groups in the exported file?"
                gridView.OptionsPrint.ExpandAllGroups = WinApplication.Messaging.GetUserChoice(message, GetMessageBoxCaption(), MessageBoxButtons.YesNo) = DialogResult.Yes
            End If
        End If
    End Sub
    Private Function HasCollapsedGroups(ByVal gridView As XafGridView) As Boolean
        If gridView.GroupCount > 0 Then
            Dim rowHandle As Integer = -1
            Do While gridView.IsValidRowHandle(rowHandle)
                If (Not gridView.GetRowExpanded(rowHandle)) Then
                    Return True
                End If
                rowHandle -= 1
            Loop
        End If
        Return False
    End Function
    Private Function GetMessageBoxCaption() As String
        Return String.Format("{0} {1}", exportController.ExportAction.Caption, exportController.ExportAction.SelectedItem)
    End Function
    Private Sub exportController_Exported(ByVal sender As Object, ByVal e As CustomExportEventArgs)
        If TypeOf e.Stream Is FileStream Then
            Dim fileName As String = (CType(e.Stream, FileStream)).Name
            If File.Exists(fileName) Then
                e.Stream.Close()
                If WinApplication.Messaging.GetUserChoice("Open the exported file?", GetMessageBoxCaption(), MessageBoxButtons.YesNo) = DialogResult.Yes Then
                    Process.Start(fileName)
                End If
            End If
        End If
    End Sub
    Protected Overrides Overloads Sub OnDeactivated()
        RemoveHandler exportController.Exporting, AddressOf exportController_Exporting
        RemoveHandler exportController.CustomExport, AddressOf exportController_CustomExport
        RemoveHandler exportController.Exported, AddressOf exportController_Exported
        MyBase.OnDeactivated()
    End Sub
End Class
End Namespace
